spinalcordtoolbox/moco/dl/inference.py

```diff
@@ -0,0 +1,180 @@
+            texture = img_raw - raw_smooth
+            out = img_warped + 1.2 * texture
+            lo, hi = np.percentile(img_raw[mask_slice > 0], [0.5, 99.5])
+            out = np.clip(out, lo, hi)
+            sharpened[..., d, t] = out
+
+    matched = np.zeros_like(sharpened)
+    for t in range(T):
+        matched[..., t] = ski_exposure.match_histograms(sharpened[..., t], mov_np[..., t])
+
+    # Save Moco output
+    base_output = add_suffix(os.path.basename(fname_data), "_mocoDL")
+    fname_moco_out = os.path.join(ofolder, base_output)
+    tmp_main = os.path.join(ofolder, "tmp_mocoDL.nii.gz")
+    nib.save(nib.Nifti1Image(matched, affine, header), tmp_main)
+    fname_moco = generate_output_file(tmp_main, fname_moco_out)
+
+    # Save mean output
+    print("[moco-dl] Computing time-averaged output volume...")
+    if mode == "dmri":
+        args = ['-i', fname_moco, '-bvec', fname_bvecs, '-a', '1', '-v', '0']
+        if fname_bvals:
+            args += ['-bval', fname_bvals]
+
+        _, fname_b0_mean, _, fname_dwi_mean = sct_dmri_separate_b0_and_dwi.main(argv=args)
+
+        generate_output_file(fname_b0_mean, add_suffix(fname_moco, "_b0_mean"))
+        generate_output_file(fname_dwi_mean, add_suffix(fname_moco, "_dwi_mean"))
+    else:
+        mean_vol = np.mean(matched, axis=3)
+        tmp_mean = os.path.join(ofolder, "tmp_mean.nii.gz")
+        nib.save(nib.Nifti1Image(mean_vol, affine, header), tmp_mean)
+        generate_output_file(tmp_mean, add_suffix(fname_moco, "_mean"))
+
+    # Save Tx, Ty, and displacement fields
+    print("[moco-dl] Saving translation params...")
+
+    Tx_img_path_tmp = os.path.join(ofolder, "tmp_Tx.nii.gz")
+    Ty_img_path_tmp = os.path.join(ofolder, "tmp_Ty.nii.gz")
+
+    nib.save(nib.Nifti1Image(Tx[np.newaxis, np.newaxis, ...], affine, header),
+             Tx_img_path_tmp)
+    nib.save(nib.Nifti1Image(Ty[np.newaxis, np.newaxis, ...], affine, header),
+             Ty_img_path_tmp)
+    generate_output_file(Tx_img_path_tmp, add_suffix(fname_moco, "_Tx"))
+    generate_output_file(Ty_img_path_tmp, add_suffix(fname_moco, "_Ty"))
+
+    # Save 5D dispfield
+    print("[moco-dl] Saving displacement fields...")
+    disp5D = np.moveaxis(flow, 0, -1)  # (H,W,D,T,3)
+    tmp_disp5D = os.path.join(ofolder, "tmp_dispfield5D.nii.gz")
+    disp_img = nib.Nifti1Image(disp5D, affine, header)
+    disp_img.header.set_intent('vector', (), '')
+    nib.save(disp_img, tmp_disp5D)
+    generate_output_file(tmp_disp5D, add_suffix(fname_moco, "_dispfield-all"))
+
+    # Save per-timepoint displacement fields
+    disp_dir = os.path.join(ofolder, "dispfield")
+    os.makedirs(disp_dir, exist_ok=True)
+
+    for t in tqdm(range(T), desc="Saving displacement fields"):
+        disp_t = disp5D[..., t, :]  # (H,W,D,3)
+        img_t = nib.Nifti1Image(disp_t, affine, header)
+        img_t.header.set_intent('vector', (), '')
+        nib.save(img_t, os.path.join(disp_dir, f"warp_t{t:04d}.nii.gz"))
```

Comment:
How long does the saving process take? I wonder if it's necessary to use `tqdm` if the operation is relatively quick...